File: android/src/main/java/expo/modules/fetcher/ExpoFetcherModule.kt
package expo.modules.fetcher

import expo.modules.fetcher.services.CookieService
import expo.modules.fetcher.services.FetchService
import expo.modules.fetcher.types.FetchInit
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoFetcherModule : Module() {
  
  private val cookieService = CookieService()
  private val fetchService = FetchService(cookieService)

  override fun definition() = ModuleDefinition {
    Name("ExpoFetcher")

    AsyncFunction("fetch") { url: String, init: FetchInit? ->
      fetchService.performFetch(url, init)
    }

    AsyncFunction("clearCookies") {
      cookieService.clearAllCookies()
    }
  }
}

File: ios/ExpoFetcherModule.swift
import ExpoModulesCore
import Foundation

public class ExpoFetcherModule: Module {
  
  private let fetchService = FetchService()
  private let legacyRequestService = LegacyRequestService()
  private let cookieService = CookieService()
  
  public func definition() -> ModuleDefinition {
    Name("ExpoFetcher")
    
    AsyncFunction("fetch") { (url: String, options: FetchInit?) -> FetchResponse in
      return try await self.fetchService.performFetch(url: url, options: options)
    }
    
    AsyncFunction("request") { (url: String, options: [String: Any]?) -> [String: Any] in
      return try await self.legacyRequestService.makeRequest(url: url, options: options ?? [:])
    }

    AsyncFunction("clearCookies") { () -> Void in
      self.cookieService.clearAllCookies()
    }
  }
}